GH-821-sub3: As a user I want type definitions of external projects separated from original project

mmews-n4 opened the pull request:
see #821

lbeurerkellner commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/utils/ProjectImportEnablingScope.java:
```diff
@@ -251,7 +256,12 @@ private IN4JSProject findProject(String projectId, IN4JSProject project) {
 		}
-		for (IN4JSProject p : project.getDependencies()) {
+
+		Iterable<IN4JSProject> dependencies = TypeDefinitionsAwareDependenciesSupplier.get(project);
```
Maybe we can make use of `IResourceScopeCache` so this comparatively expensive operation is only performed once per resource (at least in the scoping).

mmews-n4 replied:
Thanks! Yes, I introduced the `MultiCleartriggerCache` to take care of this.

lbeurerkellner commented on plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend:
```diff
@@ -330,7 +331,11 @@ class ModuleWrappingTransformation extends Transformation {
-		val depProject = n4jsCore.findProject(module.eResource.URI).orNull
+		var depProject = n4jsCore.findProject(module.eResource.URI).orNull
+		if (depProject !== null && depProject.projectType === ProjectType.DEFINITION && depProject.definesPackage !== null) {
+			depProject = n4jsCore.findAllProjectMappings.get(depProject.definesPackage);
```
The results of `n4jsCore.findAllProjectMappings` could also be cached as part of the `TranspilerState`. This would assure that we only compute the map once per transpiled resource.

mmews-n4 replied:
Thanks, I also think we should do this in a follow-up sub-task or task!

mor-n4 commented on plugins/org.eclipse.n4js.utils/src/org/eclipse/n4js/utils/MultiCleartriggerCache.java:
```diff
@@ -0,0 +1,119 @@
+ * The {@link MultiCleartriggerCache} is similar to {@link IResourceScopeCache} but provides the means to add more than
+ * one clear-trigger. A clear-trigger is an {@link URI} that will cause the cached value to be deleted.
+ */
+@Singleton
```
as discussed: because it's a singleton, better move to god bundle

mmews-n4 commented on plugins/org.eclipse.n4js.utils/src/org/eclipse/n4js/utils/MultiCleartriggerCache.java:
```diff
@@ -0,0 +1,119 @@
+			}
+		}
+		@SuppressWarnings("unchecked")
+		Entry entry = (Entry) entryMap.get(reference);
```
The problem was that it needs to be bound in the ContributingModule which does not support to bind type arguments.

mmews-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/utils/ProjectImportEnablingScope.java:
```diff
@@ -87,8 +88,10 @@
 	 */
-	public static IScope create(IN4JSCore n4jsCore, Resource resource, Optional<ImportDeclaration> importDecl,
+	public static IScope create(IN4JSCore n4jsCore, N4JSModel n4jsModel, Resource resource,
+			Optional<ImportDeclaration> importDecl,
```
I like the idea of a clear structure. However, currently the model is injected at 8 places in our codebase. Maybe we should perform a refactoring so that the model is encapsulated by the core? In this case, the model could even be of package local accessibility and/or not use injection.

mmews-n4 commented:
regarding "IN4JSProject#getDefinesPackage() should return an IN4JSProject"

I started to refactor this and needed to move the findAllProjectMapping (maps name to project) to the N4JSModel which was not easily possible. Hence I reverted the refactoring. I hope this can be done later. Anyhow, I renamed the method to getDefinesPackageName in anticipation of getDefinesPackage which would then return a project.
